### PR TITLE
Display update tag soon after save

### PR DESCRIPTION
**Pull request by rmweir**
Problem: When editing a DO cluster, the update label in the UI does not appear for approximately 30 seconds. In some scenarios it does not appear at all.

Solution: Push update to kubernetes immediately with intended changes once we know changes are to take place.

Issue: https://github.com/rancher/rancher/issues/17865

### CODE REVIEW

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 1]
```diff
@@ -287,6 +287,27 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
 		apiEndpoint, serviceAccountToken, caCert, err = p.driverUpdate(cluster, *spec)
 	} else {
 		logrus.Infof("Updating cluster [%s]", cluster.Name)
+
+		updating := false
+		for i := 0; i < 4; i++ {
+			if cluster, err = p.Clusters.Get(cluster.Name, metav1.GetOptions{}); err != nil {
```
Comment (kinarashah) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: we should first try to perform an update, if it fails and is a conflict error, only then Get object

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 2]
```diff
@@ -287,6 +287,27 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
...
+				logrus.Infof("Failed to get cluster, retrying")
+				time.Sleep(1 * time.Second)
+				continue
+			}
+			v3.ClusterConditionUpdated.Unknown(cluster)
+			if _, err = p.Clusters.Update(cluster); err != nil {
```
Comment (kinarashah) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: continue only if apierrors.IsConflict 

can refer to this example if required, https://github.com/rancher/rancher/blob/master/pkg/controllers/management/multiclusterapp/controller_multiclusterapp.go#L283

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 3]
```diff
@@ -287,6 +287,27 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
...
+				logrus.Infof("Failed to update cluster, retrying")
+				time.Sleep(1 * time.Second)
+				continue
+			} else {
+				updating = true
```
Comment (kinarashah) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: we continue in `if`, no need to add else clause, directly set updating

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 4]
```diff
@@ -287,6 +287,27 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
...
+			}
+		}
+		if !updating {
+			logrus.Infof("Failed to update cluster")
```
Comment (kinarashah) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: let's add cluster name to all log statements similar to L289

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 5]
```diff
@@ -287,6 +288,34 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
 		apiEndpoint, serviceAccountToken, caCert, err = p.driverUpdate(cluster, *spec)
 	} else {
 		logrus.Infof("Updating cluster [%s]", cluster.Name)
+
+		v3.ClusterConditionUpdated.Unknown(cluster)
+		cluster, err := p.Clusters.Update(cluster)
+		if err != nil && apierrors.IsConflict(err) {
+			for i := 0; i < 4; i++ {
+				if cluster, err = p.Clusters.Get(cluster.Name, metav1.GetOptions{}); err != nil {
+					if apierrors.IsConflict(err) {
```
Comment (kinarashah) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: we won't get Conflict error on API GET, we should just return error in this case

**Comment by dramich** [comment]
LGTM

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 6]
```diff
@@ -347,6 +351,41 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
 	return cluster, nil
 }
 
+func (p *Provisioner) setClusterStatusUpdating(cluster *v3.Cluster) {
+	v3.ClusterConditionUpdated.Unknown(cluster)
+	_, err := p.Clusters.Update(cluster)
+	if err != nil {
+		if !apierrors.IsConflict(err) {
+			logrus.Errorf("Failed to update cluster status [%s]: %v", cluster.Name, err)
```
Comment (cjellick) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: should probably log this at warn level

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 7]
```diff
@@ -347,6 +351,41 @@ func (p *Provisioner) reconcileCluster(cluster *v3.Cluster, create bool) (*v3.Cl
...
+				if _, err = p.Clusters.Update(newCluster); err != nil {
+					if apierrors.IsConflict(err) {
+						return false, nil
+					}
+					return false, fmt.Errorf("Failed to update cluster status [%s]: %v", cluster.Name, err)
+				}
+				return true, nil
+			})
+			if err != nil {
+				logrus.Errorf("%v", err)
```
Comment (cjellick) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: logging err with no additional contextual message is a bad pattern. Add some context. In this case of never getting true returned, you would get  a very generic WaitTimeoutExceeded error.
Also, should probably be logged at WARN level

**Pull request merged**